feat(data-table): sync with @carbon/react v11

IgnacioBecerra opened the pull request:
### Related Ticket(s)
#10107 

### Description
This PR syncs the Data Table component from Carbon React v11. 

In the old Carbon Web Components, the DataTable component was pretty much featureless. The Storybook features were actually implemented as a "Demo" component, which meant the adopters needed to implement every good feature themselves, instead of having the functionality already baked into the component.

For this sync effort, I made sure to keep ease-of-use as a priority, which meant that most of the features are built-in to reduce the adopter's workload when using the component. It is still customizable to some extend, mainly the filtering and sorting logic, but that's optional overall. 

Code-wise, I am handling mostly everything within the `cds-table` component to reduce the amount of code scattered throughout the rest of the files. Any properties that also affect children components get handled within the `updated()` function.

## ✅ ✅  Component work is completed, and is ready to be reviewed. Disregard earlier message. ✅ ✅ 

<del>
## ⚠️ ⚠️  Important ⚠️  ⚠️ 
This component has a lot of moving parts, with many features that can be "stackable". This might cause weird interactions but mostly **visual bugs.** Some of these have to do with very unique cases and conditions for the rows. Feature wise, this component should be fully working, but it wouldn't surprise me if there's some edge case I might've missed when testing, same with the styles.

I have opened this PR, not as a draft, since it's technically ready for visual and feature review. There are still some things I need to do, which I'll be checking off one by one, but they're mostly technical/cleaning up. I'd appreciate reviews in the meantime, I've been looking at this for so long that I definitely need another pair of eyes to double check everything's good.

- [x] Documentation updates
- [x] General code/file cleanup
- [x] Stylesheets cleanup
- [x] Reorganize directory structure
- [x] **Size styles** (they are off _sometimes_ depending on the active features, so please lmk regardless so I know where to look properly)
- [x] Skeleton story

Also small disclaimer, the `stickyHeader` feature is still experimental on v11, so expect funny looking style shenanigans here as well.

</del>

### Changelog
**New**

- `<cds-table>` now has built in features such as select, filtering, sorting, downloading, etc.
- new stories showing off new features

**Changed**

- `<cds-checkbox>`, `<cds-button>`, and `<cds-radio-button>` now have data table specific styles
- updated documentation and code sandbox example
- the expandable table variation is now easier to use without
- reduced the component bloat

**Removed**

- `<cds-table-header-cell-skeleton>`
- `<cds-table-header-expand-row>`
- `<cds-table-expand-row>`
- old stylesheets, testing, and stories

<!-- React and Web Component deploy previews are enabled by default. -->
<!-- To enable additional available deploy previews, apply the following -->
<!-- labels for the corresponding package: -->
<!-- *** "test: e2e": Codesandbox examples and e2e integration tests -->
<!-- *** "package: services": Services -->
<!-- *** "package: utilities": Utilities -->
<!-- *** "RTL": React / Web Components (RTL) -->
<!-- *** "feature flag": React / Web Components (experimental) -->

ariellalgilmore commented:
will take a look! nice!! looks like things are failing right now because of: https://github.com/carbon-design-system/carbon-for-ibm-dotcom/actions/runs/5095639427/jobs/9160766825?pr=10516#step:5:763

Also, we don't need to include anything experimental! We decided that we would only include features once they are out of the experimental phase

ariellalgilmore commented on packages/carbon-web-components/src/components/data-table/table-cell.ts:
```diff
@@ -27,10 +38,25 @@ class BXTableCell extends LitElement {
   render() {
-    return html` <slot></slot> `;
+    const { disabled, textContent } = this;
+
+    const statusText = ['Starting', 'Active', 'Disabled'];
+    const isStatus = statusText.some((word) => textContent?.includes(word));
+    const isDisabled = textContent!.includes('Disabled');
+    const cellClasses = classMap({
+      [`${prefix}--link`]: isStatus,
+      [`${prefix}--link--disabled`]: disabled || isDisabled,
+    });
```
i'm not seeing this in react https://github.com/carbon-design-system/carbon/blob/main/packages/react/src/components/DataTable/TableCell.tsx.... might help with the styles rich pointed out

ariellalgilmore replied:
i think the user should be able to use `<cds-link>` here like: https://github.com/carbon-design-system/carbon/blob/33b35df6a77d08d9c25dbf7229b0d7a0feb653f2/packages/react/src/components/DataTable/stories/shared.js#L19. and then this component can stick to just being a simple `slot`

ariellalgilmore reviewed (commented):
Looking good! Small things, to maybe note and not necessarily fix 

Batch actions story: 
- The batch actions focus should have a thicker focus... the cancel button looks good though!
![Screenshot 2023-05-31 at 10 48 13 AM](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/20210594/b88b1e5f-c962-4ec7-b2b7-5d97aec3079c)<img width="466" alt="Screenshot 2023-05-31 at 10 49 32 AM" src="https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/20210594/36d8093e-3466-4084-a8a5-2d6338cea658">

- the + and save button are missing tooltips  <img width="125" alt="Screenshot 2023-05-31 at 10 51 25 AM" src="https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/20210594/2dc28c6f-8c44-4c9a-b2b7-add5616ef1be">
- the cancel button  and search are not triggering

Dynamic:
- adding a new row missing checkbox ![Screenshot 2023-05-31 at 10 56 31 AM](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/20210594/67a04b28-26d1-4e23-b7e1-bab219c6825f)

Toolbar:
- overflow menu tooltip showing weird red box
- overflow menu showing persistenly instead of only on hover like in react
![Screenshot 2023-05-31 at 11 01 17 AM](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/20210594/e53c19c5-cb3e-4248-94b6-4ea009f8b23e)
<img width="1188" alt="Screenshot 2023-05-31 at 11 02 03 AM" src="https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/20210594/f29f2649-fd1e-4800-8354-83156013c6e1">

IgnacioBecerra commented:
Thanks for the review, Rich and Ari!

@RichKummer I've managed to sort most of the issues out, except the button size and alignment in the Filtering story. The difference in alignment comes from the Button's XL styles being different than the other sizes, but the components aren't being affected due to what I think might be a bug where the size prop isn't being passed to the toolbar content in this `Filtering` data table variation. In stories like Dynamic, you can see the toolbar contents being properly resized.

@ariellalgilmore I fixed everything except the Overflow bugs, I can't seem to replicate the issues on my end. There's no red outline showing, and the tooltip disappears after I stop hovering (after like half a second, but I think that's just the new tooltip duration thing, it can be seen on React too). I'll keep an eye on this, maybe it's a weird interaction I can't recreate for the moment.

I've added the Skeleton variant, as well as fixing the different size styles. Working on the rest of the documentation and style cleanup now, but the component itself should have all its features and styles. As for the `stickyHeader`, the functionality is sorta there (still wonky cuz its experimental and all), but I'm leaning towards removing the knob in the Playground stories. Thoughts? Maybe I can just comment out the actual feature for now instead, if anything. Lmk your thoughts.

ariellalgilmore commented:
@IgnacioBecerra! great will take a look! i think my comment was unclear for the overflow-menu....in wc all of the overflow-menus show on each row right away, but in react you only see the overflow-menu appear once you hover on a row. Also, agree on just commenting out the stickyHeader for now until it gets fully introduced into react

ariellalgilmore commented on packages/carbon-web-components/src/components/data-table/_table-action.scss:
```diff
@@ -71,44 +58,74 @@
+  &[size='xs'],
+  &[size='sm'] {
+    svg {
+      left: 0.5rem;
```
```suggestion
      left: $spacing-03;
```

ariellalgilmore commented on packages/carbon-web-components/src/components/data-table/_table-core.scss:
```diff
@@ -28,20 +28,76 @@
+    ::slotted(#{$prefix}-table-body) {
+      display: flex;
+      flex-direction: column;
+      // overflow-x: scroll;
```
```suggestion
```

ariellalgilmore reviewed (commented):
Looks good! just a few small comments and seems like the default 
```
td:not(table td) {
    padding: 1px;
}
```
is causing some extra px to appear on hover
<img width="997" alt="Screenshot 2023-06-06 at 9 06 34 AM" src="https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/20210594/44c8c6b0-e4ba-43b1-9cf3-9665586014f7">

<img width="1728" alt="Screenshot 2023-06-06 at 9 00 23 AM" src="https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/20210594/c1ef3138-6b43-4dda-9f9d-763cd3dfc132">

IgnacioBecerra commented:
@ariellalgilmore Thanks for the comments! As for the `td` thing, it's weird, I'm not getting those `td` styles locally, but just to be safe I'm adding `padding:0` if the row hasn't been expanded yet. 👍

RichKummer commented:
Hey @IgnacioBecerra noticed a few things:

1. For all variants, the labels are offset from the column content when set to `xs`, `sm`, or `md` size. It seems like it aligns fine on anything larger.
<img width="948" alt="Screenshot 2023-06-07 at 11 32 28 AM" src="https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/15643582/782bf1b2-f8bc-4efc-9722-fb5a56f55737">

2. I think you may have addressed this in your comment above, but when size is set to `xs`, the button label drops down on the button. If this is the issue you mentioned above, ignore it for this PR. Just want to confirm!

<img width="961" alt="Screenshot 2023-06-07 at 11 41 59 AM" src="https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/15643582/02c9ca8c-86a1-419b-aecc-174aae961b65">

3. For Selected, when setting the size to `xl`, the column headers move to the bottom of the container when `isSortable` is set to true.
<img width="1228" alt="Screenshot 2023-06-07 at 12 14 48 PM" src="https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/15643582/bec59b69-5542-44e7-ac81-8c7316ff351c">

4. Also for Selected, when the size is anything but `lg`, the spacing from the edge to the checkbox is rendering as `6px`, but should be `16px` throughout.
<img width="1317" alt="Screenshot 2023-06-07 at 12 19 09 PM" src="https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/15643582/cca41a61-6968-4275-97f7-24509e40164c">

5. For Toolbar, saw the same spacing issue as #4 above, similarly fine on `lg`

IgnacioBecerra commented:
@RichKummer Address your comments, except # 3. I'm not entirely sure why `vertical-align: top` isn't working (and some other variations of it), but maybe we should open a bug for it to track it in the future? It might be due to some shadow dom shenanigans due to the button styles. I'll try some other things for now, but just fyi 🤔

RichKummer commented:
Hey @IgnacioBecerra , just a few last items:

1. For the **Dynamic** version, when set to `Is sortable`, the heading does not align with the column data text. This was the only one that was still off from last time.
<img width="1306" alt="Screenshot 2023-06-13 at 10 42 04 AM" src="https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/15643582/c15af418-8d31-4f79-ac46-14207503da5d">

2. For the **Toolbar** version, when set to `xs` size and with `Overflow menu on hover`, hovering over the overflow menu makes the menu pop out of the row. In the react version, the menu button seems to be properly sized and contained in the row.

**React version**
<img width="1280" alt="Screenshot 2023-06-13 at 10 46 39 AM" src="https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/15643582/32b5f123-e818-462b-aa95-8862b33a8bb8">

**This build**
<img width="1283" alt="Screenshot 2023-06-13 at 10 47 04 AM" src="https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/15643582/ad71b5c5-7d5d-4057-aa40-a3e84d9b931d">

Everything else looks ironed out!

IgnacioBecerra commented:
@RichKummer addressed the comments!

RichKummer reviewed (approved):
LGTM! Awesome job @IgnacioBecerra !

ariellalgilmore commented on packages/carbon-web-components/src/components/data-table/data-table.scss:
```diff
@@ -16,9 +17,9 @@ $css--plex: true !default;
 @use '@carbon/styles/scss/components/checkbox';
```
Anna and i discovered we need to add missing imports for components that have styles that use `field`, `layer`, etc... i'm guessing data-table will need it. should be able to just add it to `:host(cds-table)` like : https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/10567/files#diff-61c6750e722e28b6c309405d2b52d53ede9b314acd21cf95c50130825a71c92c
